File: VB/DevExpress.XtraReports.CustomControls/XRDataContainerControlPresenters.vb
Imports System
Imports System.Drawing
Imports System.Reflection
Imports DevExpress.Drawing
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraPrinting.Native
Imports DevExpress.XtraReports.Native.Presenters
Imports DevExpress.XtraReports.Native.Printing
Imports DevExpress.XtraReports.UI

Namespace DevExpress.XtraReports.CustomControls
	Friend Class XRDataContainerControlPresenter
		Inherits ControlPresenter

		Public Sub New(ByVal control As XRDataContainerControl)
			MyBase.New(control)
		End Sub

		Public Overrides Function CreateBrick(ByVal childrenBricks() As VisualBrick) As VisualBrick
			Return New SubreportBrick(ContainerControl)
		End Function

		Protected Overridable Function CreateBrickStyle(ByVal control As XRDataContainerControl, ByVal parentBrick As VisualBrick, ByVal valueBrick As VisualBrick, ByVal record As XRDataRecord, ByVal fieldIndex As Integer, ByVal isHeader As Boolean) As BrickStyle
			Dim style As BrickStyle = GetActualBrickStyle(CType(parentBrick, DataContainerBrick), isHeader)

			If isHeader Then
				Dim printCellArgs As New PrintCellEventArgs(control.VisibleHeaders(fieldIndex), valueBrick, style)
				ContainerControl.OnPrintHeaderCell(printCellArgs)
				DirectCast(valueBrick, IDataCellBrick).CellPosition = DirectCast(valueBrick, IDataCellBrick).CellPosition Or XRDataCellPosition.Header
			Else
				Dim printCellArgs As New PrintRecordCellEventArgs(record, control.VisibleHeaders(fieldIndex), valueBrick, style)
				ContainerControl.OnPrintRecordCell(printCellArgs)
			End If

			Return style
		End Function

		Protected Overridable Function CreateCellBrick(ByVal control As XRDataContainerControl, ByVal parentBrick As VisualBrick, ByVal record As XRDataRecord, ByVal fieldIndex As Integer, ByVal isHeader As Boolean) As VisualBrick
			Dim valueBrick As VisualBrick

			Dim absoluteIndex As Integer = fieldIndex
			If Not isHeader AndAlso Not IsDesignMode Then
				Dim header As XRFieldHeader = control.VisibleHeaders(fieldIndex)
				absoluteIndex = control.Headers.IndexOf(header)
			End If

			Dim value As Object = record(absoluteIndex)

			If TypeOf value Is Boolean Then
				valueBrick = New DataCellCheckBrick(ContainerControl)
				CType(valueBrick, DataCellCheckBrick).Checked = Convert.ToBoolean(value)
			Else
				valueBrick = New DataCellTextBrick(ContainerControl)
				valueBrick.Text = Convert.ToString(value)
			End If

			valueBrick.TextValue = value

			If fieldIndex = 0 Then
				DirectCast(valueBrick, IDataCellBrick).CellPosition = DirectCast(valueBrick, IDataCellBrick).CellPosition Or XRDataCellPosition.LeftMost
			End If
			If fieldIndex = control.VisibleHeaders.Count - 1 Then
				DirectCast(valueBrick, IDataCellBrick).CellPosition = DirectCast(valueBrick, IDataCellBrick).CellPosition Or XRDataCellPosition.RightMost
			End If

			Return valueBrick
		End Function

		Protected Sub CreateDetails(ByVal parentBrick As PanelBrick, ByRef actualHeight As Single)
			LoadData()
			Dim actualCollection As XRDataRecordCollection = GetActualDataCollection()

			For i As Integer = 0 To actualCollection.Count - 1
				CreateRecordBrick(actualCollection(i), parentBrick, actualHeight)
			Next i
		End Sub

		Protected Overridable Sub CreateHeaders(ByVal parentBrick As PanelBrick, ByRef actualHeight As Single)
		End Sub

		Protected Overridable Sub CreateRecordBrick(ByVal currentRecord As XRDataRecord, ByVal parentBrick As PanelBrick, ByRef actualHeight As Single)
			If IsDesignMode OrElse (TypeOf parentBrick Is DataContainerBrick AndAlso Not DirectCast(parentBrick, DataContainerBrick).IsHeader) Then
				CreateBricksForRecord(currentRecord, parentBrick, False, actualHeight)
			End If
		End Sub

		Protected Overridable Sub CreateBricksForRecord(ByVal record As XRDataRecord, ByVal parentBrick As PanelBrick, ByVal isHeader As Boolean, ByRef actualHeight As Single)
		End Sub

		Protected Overridable Function GetActualDataCollection() As XRDataRecordCollection
			Return ContainerControl.Records
		End Function

		Protected Overridable Function GetActualBrickStyle(ByVal parentBrick As DataContainerBrick, ByVal isHeader As Boolean) As BrickStyle
			Dim resultingStyle As XRControlStyle


			If isHeader Then
				resultingStyle = New XRControlStyle(ContainerControl.fDefaultHeaderStyle)
				If CType(ContainerControl.Styles, XRDataContainerStyles).HeaderStyle IsNot Nothing Then
					ApplyStyleProperties(CType(ContainerControl.Styles, XRDataContainerStyles).HeaderStyle, resultingStyle)
				End If
			Else
				resultingStyle = New XRControlStyle(ContainerControl.fDefaultCellStyle)
				If CType(ContainerControl.Styles, XRDataContainerStyles).CellStyle IsNot Nothing Then
					ApplyStyleProperties(CType(ContainerControl.Styles, XRDataContainerStyles).CellStyle, resultingStyle)
				End If

				If parentBrick.PrintCache.RecordsCache.Count Mod 2 = 0 AndAlso CType(ContainerControl.Styles, XRDataContainerStyles).OddCellStyle IsNot Nothing Then
					ApplyStyleProperties(CType(ContainerControl.Styles, XRDataContainerStyles).OddCellStyle, resultingStyle)
				End If

				If parentBrick.PrintCache.RecordsCache.Count Mod 2 <> 0 AndAlso CType(ContainerControl.Styles, XRDataContainerStyles).EvenCellStyle IsNot Nothing Then
					ApplyStyleProperties(CType(ContainerControl.Styles, XRDataContainerStyles).EvenCellStyle, resultingStyle)
				End If
			End If

			resultingStyle.StringFormat = BrickStringFormat.Create(resultingStyle.TextAlignment, ContainerControl.WordWrap)

			Return resultingStyle
		End Function

		Private Function GetStyleProperties(ByVal style As BrickStyle) As StyleProperty
			'Had to use Reflection
			Dim pi As PropertyInfo = GetType(BrickStyle).GetProperty("SetProperties", BindingFlags.NonPublic Or BindingFlags.Instance)
			Return DirectCast(pi.GetValue(style, Nothing), StyleProperty)
		End Function

		Private Sub ApplyStyleProperties(ByVal sourceStyle As XRControlStyle, ByVal destStyle As XRControlStyle)
			Dim assignedProperties As StyleProperty = GetStyleProperties(sourceStyle)
			ApplyProperties(sourceStyle, destStyle, assignedProperties)
		End Sub

		Private Sub ApplyProperties(ByVal sourceStyle As XRControlStyle, ByVal destStyle As XRControlStyle, ByVal assignedProperties As StyleProperty)
			If (assignedProperties And StyleProperty.BackColor) <> 0 Then
				destStyle.BackColor = sourceStyle.BackColor
			End If
			If (assignedProperties And StyleProperty.BorderColor) <> 0 Then
				destStyle.BorderColor = sourceStyle.BorderColor
			End If
			If (assignedProperties And StyleProperty.BorderDashStyle) <> 0 Then
				destStyle.BorderDashStyle = sourceStyle.BorderDashStyle
			End If
			If (assignedProperties And StyleProperty.Borders) <> 0 Then
				destStyle.Borders = sourceStyle.Borders
			End If
			If (assignedProperties And StyleProperty.BorderWidth) <> 0 Then
				destStyle.BorderWidth = sourceStyle.BorderWidth
			End If
			If (assignedProperties And StyleProperty.Font) <> 0 Then
				destStyle.Font = sourceStyle.Font
			End If
			If (assignedProperties And StyleProperty.ForeColor) <> 0 Then
				destStyle.ForeColor = sourceStyle.ForeColor
			End If
			If (assignedProperties And StyleProperty.Padding) <> 0 Then
				destStyle.Padding = sourceStyle.Padding
			End If
			If (assignedProperties And StyleProperty.TextAlignment) <> 0 Then
				destStyle.TextAlignment = sourceStyle.TextAlignment
			End If
		End Sub

		Protected Overridable Sub LoadData()
			If Not IsDesignMode Then
				ContainerControl.LoadData()
			End If
		End Sub

		Protected Function MeasureTextSize(ByVal text As String, ByVal width As Single, ByVal dpi As Single, ByVal style As BrickStyle, ByVal ps As PrintingSystemBase) As Size
			Dim convertedWidth As Single = GraphicsUnitConverter.Convert(width, dpi, GraphicsDpi.Document)
			style.Padding.DeflateWidth(convertedWidth)

			ps.Graph.PageUnit = DXGraphicsUnit.Document
			ps.Graph.Font = style.Font
			Dim size As SizeF = ps.Graph.MeasureString(text, CInt(Math.Truncate(convertedWidth)), style.StringFormat.Value)
			size.Height += GraphicsUnitConverter.Convert(CSng(2F), GraphicsDpi.Pixel, GraphicsDpi.Document)

			Return System.Drawing.Size.Ceiling(GraphicsUnitConverter.Convert(style.Padding.Inflate(size, GraphicsDpi.Document), GraphicsDpi.Document, dpi))
		End Function

		Public Overrides Sub PutStateToBrick(ByVal brick As VisualBrick, ByVal ps As PrintingSystemBase)
			If TypeOf brick Is PanelBrick Then
				Dim actualHeight As Single = 0F

				brick.PrintingSystem = ps

				CreateHeaders(CType(brick, PanelBrick), actualHeight)
				CreateDetails(CType(brick, PanelBrick), actualHeight)

				CorrectBrickBounds(brick, actualHeight)
			End If
		End Sub

		Protected Overridable Sub CorrectBrickBounds(ByVal brick As VisualBrick, ByVal actualHeight As Single)
			Dim actualBounds As RectangleF

			If IsDesignMode Then
				actualBounds = ContainerControl.BoundsF
			Else
				actualBounds = New RectangleF(0, 0, ContainerControl.BoundsF.Width, actualHeight)
			End If

			VisualBrickHelper.SetBrickBounds(brick, actualBounds, ContainerControl.Dpi)
		End Sub

		Protected ReadOnly Property ContainerControl() As XRDataContainerControl
			Get
				Return CType(MyBase.control, XRDataContainerControl)
			End Get
		End Property

		Protected Overridable ReadOnly Property IsDesignMode() As Boolean
			Get
				Return False
			End Get
		End Property
	End Class
End Namespace

File: CS/DevExpress.XtraReports.CustomControls/XRDataContainerControlPresenters.cs
using System;
using System.Drawing;
using System.Reflection;
using DevExpress.Drawing;
using DevExpress.XtraPrinting;
using DevExpress.XtraPrinting.Native;
using DevExpress.XtraReports.Native.Presenters;
using DevExpress.XtraReports.Native.Printing;
using DevExpress.XtraReports.UI;

namespace DevExpress.XtraReports.CustomControls {
    internal class XRDataContainerControlPresenter : ControlPresenter
    {
        public XRDataContainerControlPresenter(XRDataContainerControl control)
            : base(control)
        {
        }

        public override VisualBrick CreateBrick(VisualBrick[] childrenBricks)
        {
            return new SubreportBrick(ContainerControl);
        }

        protected virtual BrickStyle CreateBrickStyle(XRDataContainerControl control, VisualBrick parentBrick, VisualBrick valueBrick, XRDataRecord record, int fieldIndex, bool isHeader) 
        {
            BrickStyle style = GetActualBrickStyle((DataContainerBrick)parentBrick, isHeader);

            if (isHeader)
            {
                PrintCellEventArgs printCellArgs = new PrintCellEventArgs(control.VisibleHeaders[fieldIndex], valueBrick, style);
                ContainerControl.OnPrintHeaderCell(printCellArgs);
                ((IDataCellBrick)valueBrick).CellPosition |= XRDataCellPosition.Header;
            }
            else
            {
                PrintRecordCellEventArgs printCellArgs = new PrintRecordCellEventArgs(record, control.VisibleHeaders[fieldIndex], valueBrick, style);
                ContainerControl.OnPrintRecordCell(printCellArgs);
            }

            return style;
        }

        protected virtual VisualBrick CreateCellBrick(XRDataContainerControl control, VisualBrick parentBrick, XRDataRecord record, int fieldIndex, bool isHeader) 
        {
            VisualBrick valueBrick;

            int absoluteIndex = fieldIndex;
            if (!isHeader && !IsDesignMode)
            {
                XRFieldHeader header = control.VisibleHeaders[fieldIndex];
                absoluteIndex = control.Headers.IndexOf(header);
            }

            object value = record[absoluteIndex];

            if (value is bool)
            {
                valueBrick = new DataCellCheckBrick(ContainerControl);
                ((DataCellCheckBrick)valueBrick).Checked = Convert.ToBoolean(value);
            }
            else
            {
                valueBrick = new DataCellTextBrick(ContainerControl);
                valueBrick.Text = Convert.ToString(value);
            }

            valueBrick.TextValue = value;

            if (fieldIndex == 0)
                ((IDataCellBrick)valueBrick).CellPosition |= XRDataCellPosition.LeftMost;
            if (fieldIndex == control.VisibleHeaders.Count - 1)
                ((IDataCellBrick)valueBrick).CellPosition |= XRDataCellPosition.RightMost;

            return valueBrick;
        }

        protected void CreateDetails(PanelBrick parentBrick, ref float actualHeight)
        {
            LoadData();
            XRDataRecordCollection actualCollection = GetActualDataCollection();

            for (int i = 0; i < actualCollection.Count; i++)
            {
                CreateRecordBrick(actualCollection[i], parentBrick, ref actualHeight);
            }
        }

        protected virtual void CreateHeaders(PanelBrick parentBrick, ref float actualHeight) { }

        protected virtual void CreateRecordBrick(XRDataRecord currentRecord, PanelBrick parentBrick, ref float actualHeight) 
        {
            if (IsDesignMode || (parentBrick is DataContainerBrick && !((DataContainerBrick)parentBrick).IsHeader))
            {
                CreateBricksForRecord(currentRecord, parentBrick, false, ref actualHeight);
            }
        }

        protected virtual void CreateBricksForRecord(XRDataRecord record, PanelBrick parentBrick, bool isHeader, ref float actualHeight) { }

        protected virtual XRDataRecordCollection GetActualDataCollection()
        {
            return ContainerControl.Records;
        }

        protected virtual BrickStyle GetActualBrickStyle(DataContainerBrick parentBrick, bool isHeader)
        {
            XRControlStyle resultingStyle;


            if (isHeader)
            {
                resultingStyle = new XRControlStyle(ContainerControl.fDefaultHeaderStyle);
                if (((XRDataContainerStyles)ContainerControl.Styles).HeaderStyle != null)
                    ApplyStyleProperties(((XRDataContainerStyles)ContainerControl.Styles).HeaderStyle, resultingStyle);
            }
            else
            {
                resultingStyle = new XRControlStyle(ContainerControl.fDefaultCellStyle);
                if (((XRDataContainerStyles)ContainerControl.Styles).CellStyle != null)
                    ApplyStyleProperties(((XRDataContainerStyles)ContainerControl.Styles).CellStyle, resultingStyle);

                if (parentBrick.PrintCache.RecordsCache.Count % 2 == 0 && ((XRDataContainerStyles)ContainerControl.Styles).OddCellStyle != null)
                    ApplyStyleProperties(((XRDataContainerStyles)ContainerControl.Styles).OddCellStyle, resultingStyle);

                if (parentBrick.PrintCache.RecordsCache.Count % 2 != 0 && ((XRDataContainerStyles)ContainerControl.Styles).EvenCellStyle != null)
                    ApplyStyleProperties(((XRDataContainerStyles)ContainerControl.Styles).EvenCellStyle, resultingStyle);
            }

            resultingStyle.StringFormat = BrickStringFormat.Create(resultingStyle.TextAlignment, ContainerControl.WordWrap);

            return resultingStyle;
        }

        private StyleProperty GetStyleProperties(BrickStyle style)
        {
            //Had to use Reflection
            PropertyInfo pi = typeof(BrickStyle).GetProperty("SetProperties", BindingFlags.NonPublic | BindingFlags.Instance);
            return (StyleProperty)pi.GetValue(style, null);
        }

        private void ApplyStyleProperties(XRControlStyle sourceStyle, XRControlStyle destStyle)
        {
            StyleProperty assignedProperties = GetStyleProperties(sourceStyle);
            ApplyProperties(sourceStyle, destStyle, assignedProperties);
        }

        private void ApplyProperties(XRControlStyle sourceStyle, XRControlStyle destStyle, StyleProperty assignedProperties) 
        {
            if((assignedProperties & StyleProperty.BackColor) != 0)
                destStyle.BackColor = sourceStyle.BackColor;
            if((assignedProperties & StyleProperty.BorderColor) != 0)
                destStyle.BorderColor = sourceStyle.BorderColor;
            if((assignedProperties & StyleProperty.BorderDashStyle) != 0)
                destStyle.BorderDashStyle = sourceStyle.BorderDashStyle;
            if((assignedProperties & StyleProperty.Borders) != 0)
                destStyle.Borders = sourceStyle.Borders;
            if((assignedProperties & StyleProperty.BorderWidth) != 0)
                destStyle.BorderWidth = sourceStyle.BorderWidth;
            if((assignedProperties & StyleProperty.Font) != 0)
                destStyle.Font = sourceStyle.Font;
            if((assignedProperties & StyleProperty.ForeColor) != 0)
                destStyle.ForeColor = sourceStyle.ForeColor;
            if((assignedProperties & StyleProperty.Padding) != 0)
                destStyle.Padding = sourceStyle.Padding;
            if((assignedProperties & StyleProperty.TextAlignment) != 0)
                destStyle.TextAlignment = sourceStyle.TextAlignment;
        }

        protected virtual void LoadData()
        {
            if (!IsDesignMode)
                ContainerControl.LoadData();
        }

        protected Size MeasureTextSize(string text, float width, float dpi, BrickStyle style, PrintingSystemBase ps)
        {
            float convertedWidth = GraphicsUnitConverter.Convert(width, dpi, GraphicsDpi.Document);
            style.Padding.DeflateWidth(convertedWidth);

            ps.Graph.PageUnit = DXGraphicsUnit.Document;
            ps.Graph.Font = style.Font;
            SizeF size = ps.Graph.MeasureString(text, (int)convertedWidth, style.StringFormat.Value);
            size.Height += GraphicsUnitConverter.Convert((float)2f, GraphicsDpi.Pixel, GraphicsDpi.Document);

            return Size.Ceiling(GraphicsUnitConverter.Convert(style.Padding.Inflate(size, GraphicsDpi.Document), GraphicsDpi.Document, dpi));
        }

        public override void PutStateToBrick(VisualBrick brick, PrintingSystemBase ps)
        {
            if (brick is PanelBrick)
            {
                float actualHeight = 0f;

                brick.PrintingSystem = ps;

                CreateHeaders((PanelBrick)brick, ref actualHeight);
                CreateDetails((PanelBrick)brick, ref actualHeight);

                CorrectBrickBounds(brick, actualHeight);
            }
        }

        protected virtual void CorrectBrickBounds(VisualBrick brick, float actualHeight)
        {
            RectangleF actualBounds;

            if (IsDesignMode)
                actualBounds = ContainerControl.BoundsF;
            else
                actualBounds = new RectangleF(0, 0, ContainerControl.BoundsF.Width, actualHeight);

            VisualBrickHelper.SetBrickBounds(brick, actualBounds, ContainerControl.Dpi);
        }

        protected XRDataContainerControl ContainerControl
        {
            get { return (XRDataContainerControl)base.control; }
        }

        protected virtual bool IsDesignMode { get { return false; } }
    }
}
